crates/compiler/src/llvm/data_layout.rs

```diff
@@ -0,0 +1,1415 @@
+                function_pointer_spec = Some(f_ptr);
+            } else if let Ok(iw) = NativeIntegerWidths::parser().parse(part) {
+                native_integer_widths = Some(iw);
+            } else if let Ok(npa) = NonIntegralPointerAddressSpaces::parser().parse(part) {
+                nointptr_addr_spaces = Some(npa);
+            } else if part.is_empty() {
+                continue;
+            } else {
+                Err(Error::InvalidDataLayoutSpecification(
+                    layout_string.to_string(),
+                    part.to_string(),
+                ))?;
+            }
+        }
+
+        // Compute the defaults if we have not parsed anything for a given segment.
+        let endianness = endianness.unwrap_or(Endianness::Little);
+        let mangling = mangling.unwrap_or(Mangling::ELF);
+        let stack_alignment = stack_alignment.unwrap_or(0);
+        let program_address_space = program_address_space.unwrap_or(0);
+        let global_address_space = global_address_space.unwrap_or(0);
+        let alloc_address_space = alloc_address_space.unwrap_or(0);
+        let pointer_layouts = Self::pointer_specs_or_defaults(pointer_specs);
+        let integer_layouts = Self::int_specs_or_defaults(int_specs);
+        let vector_layouts = Self::vec_specs_or_defaults(vector_specs);
+        let float_layouts = Self::float_specs_or_defaults(float_specs);
+        let aggregate_layout = aggregate_spec.unwrap_or(AggregateLayout {
+            abi_alignment:       0,
+            preferred_alignment: 64,
+        });
+        let function_pointer_layout = function_pointer_spec.unwrap_or(FunctionPointerLayout {
+            ptr_type:      FunctionPointerType::Independent,
+            abi_alignment: 64,
+        });
+        let native_integer_widths = native_integer_widths.unwrap_or(NativeIntegerWidths {
+            widths: vec![32, 64],
+        });
+        let nointptr_address_spaces =
+            nointptr_addr_spaces.unwrap_or(NonIntegralPointerAddressSpaces {
+                address_spaces: Vec::new(),
+            });
+
+        // Finally we can build the data layout
+        let layout = DataLayout {
+            endianness,
+            mangling,
+            stack_alignment,
+            program_address_space,
+            global_address_space,
+            alloc_address_space,
+            pointer_layouts,
+            integer_layouts,
+            vector_layouts,
+            float_layouts,
+            aggregate_layout,
+            function_pointer_layout,
+            native_integer_widths,
+            nointptr_address_spaces,
+        };
+
+        Ok(layout)
+    }
+
+    /// Augments the parsed floating-point layout specifications with any
+    /// missing information based on the defaults for LLVM's data layout.
+    fn float_specs_or_defaults(mut specs: Vec<FloatLayout>) -> Vec<FloatLayout> {
+        // Add the default for half-precision floats
```

Comment:
would love to see:

- these rolled up into a loop, a la:

```rust
for float_size in [16, 32, 64, 128] {
    specs.push(FloatLayout { s...: float_size, a...: float_size, p...: float_size });
}
```

- the supported float / integer / vector / etc sizes each broken out into a named const